### PR TITLE
Fix incompatible overrides of overloaded methods in concrete subclasses

### LINKED ISSUE #14002: Incorrect incompatible override report with self type + overload
Given this code:
```python
from typing import overload, TypeVar, Generic

T = TypeVar("T", str, int)

class A(Generic[T]):
    @overload
    def f(self: A[int]) -> int: ...
    @overload
    def f(self: A[str]) -> str: ...
    def f(self): ...
    
class B(A[int]):
    def f(self) -> int: ...
```
https://mypy-play.net/?mypy=latest&python=3.10&gist=736b946030f3c52e14c712ce7e95464c

mypy 0.981 reports `main.py:13: error: Signature of "f" incompatible with supertype "A"`.

This is incorrect: `B` is an `A[int]`, so it doesn't matter that it doesn't implement the second overload.

### PR DESCRIPTION
**Pull request by hauntsaninja**
Fixes #14002

### CODE REVIEW

File: mypy/checker.py [hunk 1]
```diff
@@ -1868,6 +1868,19 @@ def check_method_override_for_base_with_name(
...
+                    # If we have an overload, filter to overloads that match the self type.
+                    # This avoids false positives for concrete subclasses of generic classes,
+                    # see testSelfTypeOverrideCompatibility for an example.
+                    # It's possible we might want to do this as part of bind_and_map_method
+                    filtered_items = [
+                        item
+                        for item in original_type.items
+                        if not item.arg_types or is_subtype(active_self_type, item.arg_types[0])
+                    ]
+                    if filtered_items:
```
Comment (JelleZijlstra) [inline mypy/checker.py:1882]: If there are no filtered_items, shouldn't we always treat the override as compatible? e.g. in your `F` test below, I think the override is actually compatible: the base class says nothing about what the method should return if self is an `A[bytes]`.

**Comment by hauntsaninja** [comment]
I think the mypy_primer hits are from a different bug, will fix in another PR

File: mypy/checker.py (hunk 1 above)
Comment (hauntsaninja) [reply mypy/checker.py:1882]: Yeah, I wasn't sure about this, but it seemed like a weird situation so I chose to fail. Feels sketchy to introduce a `Callable[..., Any]` ... I'll add a comment and if it ever comes up we can reconsider

**Review by JelleZijlstra — APPROVED** [review]

**Comment by hauntsaninja** [comment]
Hmm, there are still some issues involving type variables:
```
from typing import overload, TypeVar, Union

ZAT = TypeVar("ZAT", bound="ZA")

class ZA:
    @overload
    def f(self: ZAT, x: int) -> ZAT: ...
    @overload
    def f(self, x: str) -> None: ...
    @overload
    def f(self: ZAT) -> bytes: ...
    def f(*a, **kw): ...

class ZB(ZA):
    @overload
    def f(self, x: int) -> ZB: ...
    @overload
    def f(self, x: str) -> None: ...
    def f(*a, **kw): ...
```
^this correctly gives an error on master, but this PR removes the error.

Given that this fixes a real issue I might add an xfail test and merge anyway, but let me see if I can figure out a change to make this work as well

**Comment by hauntsaninja** [comment]
Okay I added the xfail test. The correct solution probably involves the `bind_self` code. I think some of the pandas errors might come back once it's fixed too

**Comment by hauntsaninja** [comment]
Let me know if you think the new false negative is not acceptable

**Review by JelleZijlstra — APPROVED** [review]
The xfail case seems fairly contrived and false negatives are better than false positives, so I think this is an improvement. Thanks!

**Pull request merged**